### ISSUE TITLE
Bug with media range requests: content-length is returning negative values

### ISSUE THREAD

**Issue by atruskie** [QutBioacoustics/baw-server#318]
This bug is affecting production systems.

Visible effects include the audio on the listen page not loading properly. The error seems to happen the first time the audio is loaded and loads fine after a page request. I ~~have worked out Chrome sends different range request headers~~ don't know why our request range code is different in the two different scenarios:

| Action | Range request header | Content-Length | Content-Range | Response |
|--------|--------------------------|------------------|------------------|------------|
| Next button click | `Range:bytes=0-`  | `Content-Length:512001`  |`Content-Range:bytes 0-512000/591513`  | `206 Partial Content` |
|                             | `Range:bytes=512001-` | `Content-Length:20` | (empty) | `502 Bad Gateway` |
| Same page refresh | `Range:bytes=0-`  | `Content-Length:591513` | `Content-Range:bytes 0-591512/591513` | `206 Partial Content` |

I can't determine why the range-request behaviour is different for the different requests.

The response from the server is a `502 Bad Gateway` response. After intense debugging, it seems some part of the ruby codebase is returning a negative content-length for the second response:
```
Content-Range: bytes 512001-79511/591513
Content-Length: -432489
```

*Importantly*: The rails application sends these response headers back without throwing an exception. Passenger freaks out and [this case](https://github.com/phusion/passenger/blob/release-5.1.1/src/agent/Core/Controller/ForwardResponse.cpp#L119) responds with a generic bad gateway response that masks the error. The only way I found to get the above header information was to change `passenger_log_level` to level `7` (that level may not be required, but it needs to be greater than `3`).

# Tasks

- Isolate a test case against our range request code
- Patch the bug
- Investigate throwing exceptions in our code so these sorts of header based errors generate email notifications in the future

**Comment by atruskie** [comment]
Update: I've worked out why the page refresh works.

On page refresh, the browser cache kicks in. The first partial content response counts as a success and must be cached successfully, even though the second part of the partial content response fails. **The cached response only contains a partial amount of the correct data - _the missing data from the second response is not included_!**

When the cache is not involved, the data is being read by the media element... the first request succeeds, while the media is playing, the second partial content request fires, and an error occurs, and the media element stops playing immediately (after about two seconds) even though most of the audio response is available.

Sudden 💡 idea 💡: this may be related to https://github.com/QutBioacoustics/baw-client/issues/281

The same behaviour is demonstrated in the chrome auto generated page for a media download link from the listen page: 
 - with cache: the audio plays successfully the second time around
 - without cache: 
   - the audio never gets past a few seconds of playback
   - the last segment of the audio never downloads (see https://gyazo.com/7ab7f4aab5d777c1a017af36dc45fffe) which perfectly explains  https://github.com/QutBioacoustics/baw-client/issues/281
  

---

This doesn't completely explain the problem. I suspect this bug is only happening when the requested media is of high enough quality to trigger partial byte range requests.

**Comment by atruskie** [comment]
Bug patched in 02a215820734de7c012b966f2a0d1598d97dbd84 - waiting for review.
